<!--
%\VignetteIndexEntry{Exploring the 1.3 million brain cell scRNA-seq data from 10X Genomics}
%\VignettePackage{BiocStyle}
%\VignetteEngine{knitr::knitr}
-->

```{r, echo=FALSE, results="hide", message=FALSE}
require(knitr)
opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
```

```{r style, echo=FALSE, results='asis'}
BiocStyle::markdown()
```

# Exploring the 1.3 million brain cell scRNA-seq data from 10X Genomics
Package: `r Biocpkg("TENxBrainData")` <br />
Author: Aaron Lun (alun@wehi.edu.au) <br />
Compilation date: `r Sys.Date()`

# Introduction

The `r Biocpkg("TENxBrainData")` package provides a R/Bioconductor resource for representing and manipulating the 1.3 million brain cell single-cell RNA-seq (scRNA-seq) data set generated by 10X Genomics.
It makes extensive use of the `r Biocpkg("HDF5Array")` package to avoid loading the entire data set in memory, instead storing the counts on disk as a HDF5 file and loading subsets of the data into memory upon request.

# Loading in the data 

We use the `TENxBrainData` function to download the relevant files from Bioconductor's ExperimentHub web resource.
This includes the HDF5 file containing the counts, as well as the metadata on the rows (genes) and columns (cells).
The output is a single `SingleCellExperiment` object from the `r Biocpkg("SingleCellExperiment")` package.
This is equivalent to a `SummarizedExperiment` class but with a number of features specific to single-cell data.


```{r}
library(TENxBrainData)
tenx <- TENxBrainData()
tenx
```

The first call to this function will take some time due to the need to download some moderately large files.
The files are then stored locally such that ensuing calls in the same or new sessions are very fast.

```{r}
TENxBrainData() # cheap
```

The count matrix itself is represented as a `DelayedMatrix` from the `r Biocpkg("DelayedArray")` package.
This wraps the underlying HDF5 file in a container that can be manipulated in R. 
Each count represents the number of unique molecular identifiers (UMIs) assigned to a particular gene in a particular cell.

```{r}
counts(tenx)
```

# Exploring the data

To quickly explore the data set, we compute some summary statistics on the count matrix.
We increase the `r Biocpkg("DelayedArray")` block size to indicate that we can use up to 2 GB of memory for loading the data into memory from disk.

```{r}
options(DelayedArray.block.size=2e9)
```

We obtain the library sizes for each cell by taking the column sums of the matrix.

```{r}
lib.sizes <- colSums(counts(tenx))
hist(log10(lib.sizes), xlab=expression(Log[10]~"Library size"), col="grey80")
```

We also compute the number of genes expressed in each cell (i.e., with non-zero counts).

```{r}
n.exprs <- colSums(counts(tenx)!=0L)
hist(n.exprs, xlab="Number of detected genes", col="grey80")
```

Similar operations are applied to the rows to obtain the average count for each gene.

```{r}
ave.exprs <- rowMeans(counts(tenx))
hist(log10(ave.exprs), xlab=expression(Log[10]~"Average count"), col="grey80")
```

We also examine the top most highly-expressing genes in this data set.

```{r}
o <- order(ave.exprs, decreasing=TRUE)
head(data.frame(Gene=rowData(tenx)$Symbol, Average=ave.exprs)[o,])
```

More advanced analysis procedures are implemented in various Bioconductor packages - see the `SingleCell` biocViews for more details.

# Session information

```{r}
sessionInfo()
```
